designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,62 @@
+- Repo: eslint/eslint
+- Start Date: 2020-06-04
+- RFC PR: 
+- Authors: Ilya Volodin
+
+# Generic AST support
+
+## Summary
+
+Allow ESLint to traverse and lint generic (non estree compliant AST) and allow parsers to specify information about AST that would drive ESLint's traversal and features.
+
+## Motivation
+
+ESLint has cemented itself as a go to tool for JavaScript/TypeScript ecosystem for style checking and linting. However, there are a lot of connected languages/technologies that are used every day by ESLint users that are either lacking linters, or require yet another tool to be setup and configured for the repository. Examples include HTML, CSS, SCSS, LESS, GraphQL, JSON, etc.
+With addition of config overrides users can now setup separate configuration per glob, so that would allow them to have a single instance of ESLint and a single configuration that would allow linting of the entire repository, including all non javascript/typescript files.
+
+## Detailed Design
+
+### Initial assumptions
+* This RFC does not expect or implies that existing core or plugin rules will work for any files that are parsed through non-estree compliant parsers. Every parser that chooses to output custom AST will have to provide new rules that will work for this AST in the form of plugins.
+* This RFC doesn't propose full feature parity for new parser with existing ESLint capabilities. For example, it doesn't propose a new way to provide custom Code Path Analyzer, or support inline directives for controlling ESLint's flow for custom ASTs. However, those functionalities can be added later with a separate RFC(s).
+
+Currently ESLint has a lot of hardcoded assumptions about AST. For example, it assumes that nodes are identified by the property called `type` and that property exists on every node. It assumes that top-most node is always `Program` and that AST supports code-path-analysis (which is very specific to JavaScript). This RFC proposes to modify expected output from the `parseForESLint` method of the custom parsers to include additional property called `parserSupportOptions` of type object with four new properties (for now):
+ - nodeIdentifier: function | undefined - Specifies a function that can retrieve the name of the ASTNode property that identifies node. If the property is not provided, ESLint will default to `type` for backwards compatibility.
```

Comment:
Thought: If this is a function, maybe it should have an action name? I'm still partial to `getNodeType()`, but I'm not opposed to something else. Also, I suppose this could be allowed to be a function or a string for simplicity?

Not critical, mostly posted here for others' thoughts.

Reply:
Sure, that sounds like a good enhancement to me, I'll make the change, since most parsers probably wouldn't need support for variable node name. In terms of name, I think `getNodeType` is a bit misleading, for me, for example, it would mean given the node, get me it's type, not the get me the property that identifies all nodes. We could change it to getNodeIdentifier(), but that just feels overly verbose to me.

Reply:
Ah, I see the confusion. To me, we need `getNodeType(node)` to return the node type string (like "Program") rather than just returning the property name. As in the TypeScript example I mentioned, there's just no way to get the node type unless you calculate it, which means we can't just return a property to go look it up.

Reply:
So what you are saying is that `getNodeType` function would take an ASTNode and return back a string representing the type of that node, because TypeScript ASTNodes do not contain identifying property? Am I understanding this correctly? My biggest concern here would be performance implication of this change, since for every single node we wold have to call into the parser. But I guess that could be optimized later. I'm also thinking that if we make `getNodeType` be either a function or a string, the name will be somewhat confusing when it's of type string. But it's a minor concern.

Reply:
Yes, that's correct. If we want to still allow a string to be specified, then we can probably keep the name as `nodeIdentifier` (or maybe `nodeType`).